File: disk_sch_java/src/SCAN.java
/*
    SCAN - Disk Scheduling Algorithm (Elevator)

    scans down towards the nearest end and then when it hits the bottom,
    it scans up servicing the requests that it didn't get going down.
    If a request comes in after it has been scanned it will not be serviced
    until the process comes back down or moves back up.

 */

import java.util.Arrays;
import java.util.Scanner;

public class SCAN {
    public static void main(String [] args){
        Scanner input = new Scanner(System.in);
        int sum =0;
        int head;
        int [] queue;
        int max;
        int dloc = 0;  //location of disk(head) in arr

        System.out.println("Input no of disk locations");
        int qSize = Integer.parseInt(input.nextLine());

        queue = new int[qSize+1];
        System.out.println("Enter head position");
        head = Integer.parseInt(input.nextLine());

        System.out.println("Input elements in disk queue");
        for(int i=0; i<qSize; i++){
            queue[i] = Integer.parseInt(input.nextLine());
        }

        queue[qSize] = head;
        qSize++;

        //sort disk locations queue
        Arrays.sort(queue);

        max = queue[qSize-1];

        for(int i = 0; i<qSize; i++){
            if(head == queue[i]){
                dloc = i;
                break;
            }
        }

        for(int i = dloc; i>=0; i--){
            System.out.print(queue[i] + "-->");
        }

        System.out.print("0-->");

        for(int i=dloc+1; i<qSize; i++){
            System.out.print(queue[i] + "-->");
        }

        System.out.println();

        sum=head+max;
        System.out.println(" Movement of total cylinders " + sum);

    }

}

File: scan.c

/*
SCAN - Disk Scheduling Algorithm (Elevator)

scans down towards the nearest end and then when it hits the bottom,
it scans up servicing the requests that it didn't get going down.
If a request comes in after it has been scanned it will not be serviced
until the process comes back down or moves back up.

*/

#include <stdio.h>
#include <stdlib.h>

#define LOW 0
#define HIGH 199


int main(){
  int queue[20];
  int head, max, q_size, temp, sum;
  int dloc; //location of disk (head) arr


  printf("%s\t", "Input no of disk locations");
  scanf("%d", &q_size);

  printf("%s\t", "Enter head position");
  scanf("%d", &head);

  printf("%s\n", "Input elements into disk queue");
  for(int i=0; i<q_size; i++){
    scanf("%d", &queue[i]);
  }

  queue[q_size] = head; //add RW head into queue
  q_size++;

  //sort the array
  for(int i=0; i<q_size;i++){
    for(int j=i; j<q_size; j++){
      if(queue[i]>queue[j]){
        temp = queue[i];
        queue[i] = queue[j];
        queue[j] = temp;
      }
    }
  }

  max = queue[q_size-1];

  //locate head in the queue
  for(int i=0; i<q_size; i++){
    if(head == queue[i]){
      dloc = i;
      break;
    }
  }

  if(abs(head-LOW) <= abs(head-HIGH)){

      for(int j=dloc; j>=0; j--){
        printf("%d --> ",queue[j]);
      }
      for(int j=dloc+1; j<q_size; j++){
        printf("%d --> ",queue[j]);
      }

      } else {

      for(int j=dloc+1; j<q_size; j++){
          printf("%d --> ",queue[j]);
      }
      for(int j=dloc; j>=0; j--){
          printf("%d --> ",queue[j]);
      }

  }




  sum  = head + max;
  printf("\nmovement of total cylinders %d", sum);

  return 0;
}
